File: src/test/java/com/dmarchesseault/sudoku/SolverStrategyScanTest.java
package com.dmarchesseault.sudoku;

import com.dmarchesseault.tools.StopWatch;
import org.junit.Test;

public class SolverStrategyScanTest
{

//    @Test
    public void test_1() throws BoardSanityException
    {
        final SolverStrategyScan solver = new SolverStrategyScan();

        final Board board = new Board();
        SudokuHelper.populateBoardFromText(board, SudokuSamples.SUDOKU_BOARD_SAMPLE_01);

        print(board, 0);

        solver.apply(board);

        print(board, 0);
    }

//    @Test
    public void test_10() throws BoardSanityException
    {
        test(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_10),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_10_SOLUTION));
    }

//    @Test
    public void test_20() throws BoardSanityException
    {
        test(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_20),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_20_SOLUTION));
    }

    @Test
    public void test_30() throws BoardSanityException
    {
        test(
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_30),
                SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_30_SOLUTION));
    }

//    @Test
    public void test_40() throws BoardSanityException
    {
        test(SudokuHelper.populateBoardFromText(new Board(), SudokuSamples.SUDOKU_BOARD_SAMPLE_40));
    }

    @Test
    public void test_scanIdenticalFlagsInGroups() throws BoardSanityException
    {
        // This board is designed to have 3 cells on the first row that have candidate flags for digits 7 and 8
        // The pair of clles that have only those two flags should be detected and the flags should be reset in the other
        // cell i the last column
        final String BOARD =
                "+-----+-----+-----+" +
                "|1 2 3|4 5 .|. . .|" +
                "|. . .|. . 7|. . .|" +
                "|. . .|. . 8|. . .|" +
                "+-----+-----+-----+" +
                "|. . .|. . .|6 . .|" +
                "|. . .|. . .|9 . .|" +
                "|. . .|. . .|. . .|" +
                "+-----+-----+-----+" +
                "|. . .|. . .|. 6 .|" +
                "|. . .|. . .|. 9 .|" +
                "|. . .|. . .|. . .|" +
                "+-----+-----+-----+";

        final Board board = new Board();
        SudokuHelper.populateBoardFromText(board, BOARD);
        test(board);
    }


    public void test(final Board board,final Board solutionBoard) throws BoardSanityException
    {
        board.setSolutionBoard(solutionBoard);
        test(board);
    }

    public void test(final Board board) throws BoardSanityException
    {
        print(board, 0);
        final SolverStrategy solverStrategy = new SolverStrategyScan();

        final StopWatch stopWatch = new StopWatch().start();
        int n;
        do {
            n = solverStrategy.apply(board);
            print(board, n);
        } while (n > 0);
        System.out.println("Solver elapsed time (ms): " + stopWatch.stop().elapsed());
    }

    private void print(final Board board, int hits)
    {
        System.out.println(hits + " hits");
        System.out.println(SudokuHelper.printBoardDigits(board, '.'));
        System.out.println(SudokuHelper.printBoardDigitFlags(board));
    }
}

File: sudoku-js/src/test/js/sudoku/SolverStrategyScanTest.js
console && console.log("### ns.SolverStrategyScanTest.js begin");

(function(ns,test,lang,tools)
{

    LOG = Logger.getLogger("SolverStrategyScanTest");
    ns.solverStrategyScanTestSuite = new test.TestSuite({name: "SolverStrategyScanTest"});

    let printBoard = function(board, hits)
    {
        if (LOG._isDebugOn) {
            LOG.debug(hits + " hits");
            LOG.debug(ns.SudokuHelper.printBoardDigits(board, '.'));
            LOG.debug(ns.SudokuHelper.printBoardDigitFlags(board));
        }
    };


    let testBoard = function (board)
    {
        const solverStrategy = new ns.SolverStrategyScan();
        printBoard(board, 0);

        const stopWatch = new tools.StopWatch().start();
        let n;
        do {
            n = solverStrategy.apply(board);
            printBoard(board, n);
        } while (n > 0);
        LOG.trace("Solver elapsed time (ms): " + stopWatch.stop().elapsed());
    };


    let testBoardSolution = function (board, solutionBoard)
    {
        board.setSolutionBoard(solutionBoard);
        testBoard(board);
    };


    ns.solverStrategyScanTestSuite.addTest(new test.TestCase({name: "test_1", active: true}, function() {
        const solver = new ns.SolverStrategyScan();

            const board = new ns.Board();
            ns.SudokuHelper.populateBoardFromText(board, ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_01);

            printBoard(board, 0);

            solver.apply(board);

            printBoard(board, 0);
        }));


    ns.solverStrategyScanTestSuite.addTest(new test.TestCase({name: "test_10", active: true},function() {
            testBoardSolution(
                    ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_10),
                    ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_10_SOLUTION));
        }));


    ns.solverStrategyScanTestSuite.addTest(new test.TestCase({name: "test_20", active: true}, function() {
        testBoardSolution(
                    ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_20),
                    ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_20_SOLUTION));
        }));


    ns.solverStrategyScanTestSuite.addTest(new test.TestCase({name: "test_30", active: true}, function() {
        testBoardSolution(
                    ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_30),
                    ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_30_SOLUTION));
        }));


    //@Test(expected=BoardSanityException.class)
    ns.solverStrategyScanTestSuite.addTest(new test.TestCase({name: "test_40", expectedExceptionType: ns.BoardSanityException.TYPE, active: true}, function() {
        testBoard(ns.SudokuHelper.populateBoardFromText(new ns.Board(), ns.SudokuSamples.SUDOKU_BOARD_SAMPLE_40));
        }));


    ns.solverStrategyScanTestSuite.addTest(new test.TestCase({name: "test_scanIdenticalFlagsInGroups", active: true}, function() {
            // This board is designed to have 3 cells on the first row that have candidate flags for digits 7 and 8
            // The pair of clles that have only those two flags should be detected and the flags should be reset in the other
            // cell i the last column
            const BOARD =
                    "+-----+-----+-----+" +
                    "|1 2 3|4 5 .|. . .|" +
                    "|. . .|. . 7|. . .|" +
                    "|. . .|. . 8|. . .|" +
                    "+-----+-----+-----+" +
                    "|. . .|. . .|6 . .|" +
                    "|. . .|. . .|9 . .|" +
                    "|. . .|. . .|. . .|" +
                    "+-----+-----+-----+" +
                    "|. . .|. . .|. 6 .|" +
                    "|. . .|. . .|. 9 .|" +
                    "|. . .|. . .|. . .|" +
                    "+-----+-----+-----+";

            const board = new ns.Board();
            ns.SudokuHelper.populateBoardFromText(board, BOARD);
            testBoard(board);
        }));

}(Module("sudoku"),Module("test"),Module("lang"),Module("tools")));

console && console.log("### ns.SolverStrategyScanTest.js end");
